urlencode.1.md
%urlencode(1) user manual | version 1.3.5 f86e208
% R. S. Doiel
% 2026-02-12

# NAME

urlencode

# SYNOPSIS

urlencode [OPTIONS] [STRING]

# DESCRIPTION

urlencode is a simple command line utility to URL encode content. By default
it reads from standard input and writes to standard out.  You can
also specifty the string to encode as a command line parameter.

You can provide the string to encode as a command line parameter otherwise it
will be read from standard input.

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

-query
: Use URL query style encoding (use plus for spaces)

-newline
: Append a trailing newline

# EXAMPLE

echo "This is the string to encode & nothing else!" | urlencode

would yield

    This%20is%20the%20string%20to%20encode%20&%20nothing%20else%0A




RELEASES.md

Release Process and tags
========================

Preleases and production releases
---------------------------------

This describes a simple release process organized around semantic versioned tags.

When a new release is ready it should be tag as v0.XX.XX-pre (where XX is a number) and published as a 'pre-release' on
Github. Send a note to the development group so someone other than the proposer of the release will be able to 
indepentently evaluate the release changes.  At a minimum they should run `bash test_cmds.bash` and walk through some 
of the documentation examples. If the release is verified as ready then a new release will be cut with a tag like
v0.XX.XX (there the XX are the same as in v0.XX.XX-pre). E.g. v0.0.10-pre would become v0.0.10 on success.

If the release fails verification, bugs need to be fixed and a new release proposed create after the fixes. 
The "patch" number persion should be incremented in v0.XX.XX-pre (e.g. v0.0.10-pre would be followed by 
v0.0.11-pre indicating that patches). 

NOTE: This means there their can be skips in the production patch numbers between release. E.g. v.0.0.9 might
be followed by v0.0.10-pre, v0.0.11-pre, v0.0.12-pre before a v0.0.12 appears as a production release.

Production and pre-releases should include Zip files of the compiled cli to be tested by `bash test_cmds.bash`.

Dev releases
------------

Dev release may happend from time to time as needed. They should always end in a '-dev' version number (e.g. v0.0.10-dev). 
They normally should not have any pre-compiled binaries to avoid confusion. They should be flagged as draft (pre-release)
on Github.

Making a release
----------------

1. Set the version number in PACKAGE.go (where PACKAGE is the name of the package, e.g. dataset is the name of the dataset
package so you'd change the version number in dataset.go).
2. Run `make clean`
3. Run `make test` and make sure they pass, if some fail document it if you plan to release it (e.g. GSheet integration not tested because...)
4. Run `make release`

You are now ready to go to Github and create a release. If you are uploading compiled versions upload the zip files in the _dist_
folder.



cmd/urlencode/urlencode.go
//
// urlencode.go is a simple command line utility to encode a string in a URL friendly way.
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
//
// Copyright (c) 2021, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"net/url"
	"path"
	"os"
	"strings"

	// CaltechLibrary Packages
	"github.com/caltechlibrary/datatools"
)

var (
	helpText = `%{app_name}(1) user manual | version {version} {release_hash}
% R. S. Doiel
% {release_date}

# NAME

{app_name}

# SYNOPSIS

{app_name} [OPTIONS] [STRING]

# DESCRIPTION

{app_name} is a simple command line utility to URL encode content. By default
it reads from standard input and writes to standard out.  You can
also specifty the string to encode as a command line parameter.

You can provide the string to encode as a command line parameter otherwise it
will be read from standard input.

# OPTIONS

-help
: display help

-license
: display license

-version
: display version

-query
: Use URL query style encoding (use plus for spaces)

-newline
: Append a trailing newline

# EXAMPLE

echo "This is the string to encode & nothing else!" | urlencode

would yield

    This%20is%20the%20string%20to%20encode%20&%20nothing%20else%0A

`
	// Standard Options
	showHelp         bool
	showLicense      bool
	showVersion      bool

	// App Options
	useQueryEscape bool
	newLine          bool
)

func main() {
	appName := path.Base(os.Args[0])
	version := datatools.Version
	license := datatools.LicenseText
	releaseDate := datatools.ReleaseDate
	releaseHash := datatools.ReleaseHash

	// Standard Options
	flag.BoolVar(&showHelp, "help", showHelp, "display help")
	flag.BoolVar(&showLicense, "license", showLicense, "display license")
	flag.BoolVar(&showVersion, "version", showVersion, "display version")
	flag.BoolVar(&newLine, "newline", false, "append a newline character")
	flag.BoolVar(&useQueryEscape, "query", false, "use query escape (pluses for spaces)")
 
	// Parse environment and options
	flag.Parse()
	args := flag.Args()

	in := os.Stdin
	out := os.Stdout
	eout := os.Stderr

	// Handle the default options
	if showHelp {
			fmt.Fprintf(out, "%s\n", datatools.FmtHelp(helpText, appName, version, releaseDate, releaseHash))
			os.Exit(0)
	}
	if showLicense == true {
			fmt.Fprintf(out, "%s\n", license)
			os.Exit(0)
	}
	if showVersion == true {
			fmt.Fprintf(out, "datatools, %s %s %s\n", appName, version, releaseHash)
			os.Exit(0)
	}

	nl := "\n"
	if newLine == false {
		nl = ""
	}

	var src string

	if len(args) > 0 {
		src = strings.Join(args, " ")
	} else {
		buf, err := ioutil.ReadAll(in)
		if err != nil {
			fmt.Fprintf(eout, "%s\n", err)
			os.Exit(1)
		}
		src = fmt.Sprintf("%s", buf)
	}
	if useQueryEscape == true {
		fmt.Fprintf(out, "%s%s", url.QueryEscape(src), nl)
	} else {
		fmt.Fprintf(out, "%s%s", url.PathEscape(src), nl)
	}
}
